### PR TITLE
Databases page

### PR DESCRIPTION
**Pull request by mxsnyder**
Changes and tests for the databases page and the locations page. The big deal here is the filter-page angular app, and the use of directives.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -15,82 +18,120 @@ const uglify = require('gulp-uglify');
...
+
+
+gulp.task('process-app-js', () => {
+	const appRootFolder = 'js/apps';
+	const appFolders = fs.readdirSync(appRootFolder).filter((file) => {
+		return fs.statSync(path.join(appRootFolder, file)).isDirectory();
+	});
+
+	appFolders.forEach((folder) => {
+		gulp.src([`js/apps/${folder}/app.js`, `js/apps/${folder}/dataServices/**/*.js`, `js/apps/${folder}/services/**/*.js`, `js/apps/${folder}/controllers/**/*.js`, `js/apps/${folder}/directives/**/*.js`])
```
Comment (ascire) [inline gulpfile.js]: TIL about template literals...

**Comment by ascire** [comment]
Are you sure you want to include package-lock.json?

File: js/apps/filter-page/controllers/filterPageCtrl.js [hunk 2]
```diff
@@ -30,7 +31,6 @@
 			self.filters = filters;
 			self.allCardData = cardData;
 			self.items = cardData;
-			$scope.$apply();
 
 			angular.element('#results-display').trigger('bcpl.filter.changed', { items: self.items });
```
Comment (martypowell) [inline js/apps/filter-page/controllers/filterPageCtrl.js]: Should you dump this into a method, and call it in setFilter and wherever else you need it. See line 16 and 19

File: js/apps/filter-page/directives/card.js [hunk 3]
```diff
@@ -1,9 +1,21 @@
-(() => {
+((app) => {
 	'use strict';
 
-	const cardDirective = ($compile, $injector) => {
+	const cardDirective = ($compile, $injector, $animate) => {
 		const cardLink = ($scope, element, attrs) => {
 			element.append($compile($injector.get(attrs.template + 'Service').template)($scope));
+
+			/* $scope.$watch('cardData', () => {
```
Comment (martypowell) [inline js/apps/filter-page/directives/card.js]: remove?

File: js/apps/filter-page/directives/tag.js [hunk 4]
```diff
@@ -3,11 +3,14 @@
...
 	const tagDirective = (tagParsingService) => {
 		const tagLink = function filterLink($scope) {
-			$scope.toggleFilter = (activeFilter, $event) => {
-				const $element = angular.element($event.currentTarget);
+			$scope.toggleFilter = (activeFilter) => {
+				const tagFamiliesForCard = tagParsingService.parseTags($scope.tagData.attributes);
+				const activeTagName = tagParsingService.extractTagName(activeFilter);
+				const activeTagFamily = tagFamiliesForCard.filter((tagFamily) => {
+					return tagFamily.tags.indexOf(activeTagName) !== -1;
+				})[0];
```
Comment (martypowell) [inline js/apps/filter-page/directives/tag.js]: Will this array ever be null?

File: js/apps/filter-page/directives/tag.js [hunk 5]
```diff
@@ -20,7 +23,7 @@
 				activeFilters: '='
 			},
 			restrict: 'E',
-			template: '<li ng-repeat="tag in tagData"><button ng-click="toggleFilter(tag, $event)" ng-class="{active: activeFilters.indexOf(extractTagName(tag)) !== -1}">{{extractTagName(tag)}}</button></li>',
+			template: '<li ng-repeat="tag in tagData.attributes"><button ng-click="toggleFilter(tag, $event)" ng-class="{active: activeFilters.indexOf(extractTagName(tag)) !== -1}">{{extractTagName(tag)}}</button></li>',
```
Comment (martypowell) [inline js/apps/filter-page/directives/tag.js]: Move to external template

File: js/apps/filter-page/directives/tag.js (hunk 4 above)
Comment (mxsnyder) [reply js/apps/filter-page/directives/tag.js]: I don't think it's possible, but I added code to protect against it if it ever happens.

File: js/apps/filter-page/controllers/filterPageCtrl.js [hunk 6]
```diff
@@ -1,11 +1,39 @@
 ((app) => {
 	'use strict';
 
-	const FilterPageCtrl = function FilterPageCtrl($scope, cardService, tagParsingService, $animate) {
+	const cardVisibilityFilter = (tagParsingService) => {
```
Comment (martypowell) [inline js/apps/filter-page/controllers/filterPageCtrl.js]: Move filter to it's own file

Comment (mxsnyder) [reply js/apps/filter-page/controllers/filterPageCtrl.js]: Done!

**Pull request merged**